src/cluster/dbscan.rs

```diff
@@ -0,0 +1,252 @@
+//! # DBSCAN Clustering
+//!
+//! DBSCAN - Density-Based Spatial Clustering of Applications with Noise.
+//!
+//! Example:
+//!
+//! ```
+//! use smartcore::linalg::naive::dense_matrix::*;
+//! use smartcore::cluster::dbscan::*;
+//! use smartcore::math::distance::Distances;
+//! use smartcore::neighbors::KNNAlgorithmName;
+//! use smartcore::dataset::generator;
+//!
+//! // Generate three blobs
+//! let blobs = generator::make_blobs(100, 2, 3);
+//! let x = DenseMatrix::from_vec(blobs.num_samples, blobs.num_features, &blobs.data);
+//! // Fit the algorithm and predict cluster labels
+//! let labels = DBSCAN::fit(&x, Distances::euclidian(), DBSCANParameters{
+//!     min_samples: 5,
+//!     eps: 3.0,
+//!     algorithm: KNNAlgorithmName::CoverTree
+//! }).and_then(|dbscan| dbscan.predict(&x));
+//!
+//! println!("{:?}", labels);
+//! ```
+//!
+//! ## References:
+//!
+//! * ["A Density-Based Algorithm for Discovering Clusters in Large Spatial Databases with Noise", Ester M., Kriegel HP., Sander J., Xu X.](http://faculty.marshall.usc.edu/gareth-james/ISL/)
+//! * ["Density-Based Clustering in Spatial Databases: The Algorithm GDBSCAN and its Applications", Sander J., Ester M., Kriegel HP., Xu X.](https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.63.1629&rep=rep1&type=pdf)
+
+extern crate rand;
+
+use std::fmt::Debug;
+use std::iter::Sum;
+
+use serde::{Deserialize, Serialize};
+
+use crate::error::Failed;
+use crate::linalg::{row_iter, Matrix};
+use crate::math::distance::Distance;
+use crate::math::num::RealNumber;
+use crate::neighbors::{KNNAlgorithm, KNNAlgorithmName};
```

Comment:
Since we use `KNNAlgorithm` and `KNNAlgorithmName` here let's move these into `crate::algorithm::neighbour`?